### PR TITLE
Add authorization endpoint

### PR DESCRIPTION
**Pull request by KirillPamPam**
Add an authorization endpoint which accepts a jwt token (from dproxy). Then this token is verified using provider's public key, some claims are also verified (`version`, `issuer` and `issAt`). If the token is correct then provider's jwt token is created  with passing `sessionId` claim. It is signed by provider's private key and sent back to dproxy side.
RSA keys are used.

### CODE REVIEW

File: src/main/kotlin/io/emeraldpay/dshackle/auth/processor/AuthProcessor.kt [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                .asException()
+    }
+}
+
+abstract class AuthProcessor {
+
+    open fun process(keyPair: KeyReader.KeyPair, token: String): String {
+        try {
+            val verifier: JWTVerifier = JWT.require(verifyingAlgorithm(keyPair.drpcPublicKey))
+                .withIssuer("drpc")
```
Comment (Termina1) [inline src/main/kotlin/io/emeraldpay/dshackle/auth/processor/AuthProcessor.kt:38]: should be configurable

**Review by Termina1 — CHANGES_REQUESTED** [review]

File: src/main/kotlin/io/emeraldpay/dshackle/auth/service/AuthService.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    private fun auth(authConfig: AuthorizationConfig, token: String): String {
+        try {
+            val keyPair = rsaKeyReader.getKeyPair(authConfig.providerPrivateKeyPath, authConfig.drpcPublicKeyPath)
+            val decodedJwt = JWT.decode(token)
+
+            return authProcessorFactory
+                .getAuthProcessor(decodedJwt)
+                .process(keyPair, token)
+        } finally {
+            lock.unlock()
```
Comment (Termina1) [inline src/main/kotlin/io/emeraldpay/dshackle/auth/service/AuthService.kt:46]: may be unlock in parent method?

File: src/main/kotlin/io/emeraldpay/dshackle/auth/service/KeyReader.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import java.security.PrivateKey
+import java.security.PublicKey
+
+interface KeyReader {
+
+    fun getKeyPair(providerPrivateKeyPath: String, drpcPublicKeyPath: String): KeyPair
+
+    data class KeyPair(
+        val providerPrivateKey: PrivateKey,
+        val drpcPublicKey: PublicKey
```
Comment (Termina1) [inline src/main/kotlin/io/emeraldpay/dshackle/auth/service/KeyReader.kt:12]: we shouldn't use DRPC naming

File: src/main/kotlin/io/emeraldpay/dshackle/auth/processor/AuthProcessor.kt [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            verifier.verify(token)
+        } catch (e: Exception) {
+            throw Status.INVALID_ARGUMENT
+                .withDescription("Invalid token: ${e.message}")
+                .asException()
+        }
+
+        return processInternal(keyPair.providerPrivateKey)
+            .also {
+                AuthContext.putTokenInContext(it)
```
Comment (Termina1) [inline src/main/kotlin/io/emeraldpay/dshackle/auth/processor/AuthProcessor.kt:47]: It's very strange place to put something in context, we should do this in AuthService may be?

File: src/main/kotlin/io/emeraldpay/dshackle/auth/processor/AuthProcessor.kt (hunk 1 above)
Comment (KirillPamPam) [reply src/main/kotlin/io/emeraldpay/dshackle/auth/processor/AuthProcessor.kt]: done

File: src/main/kotlin/io/emeraldpay/dshackle/auth/service/AuthService.kt (hunk 2 above)
Comment (KirillPamPam) [reply src/main/kotlin/io/emeraldpay/dshackle/auth/service/AuthService.kt]: Removed `lock` at all

File: src/main/kotlin/io/emeraldpay/dshackle/auth/service/KeyReader.kt (hunk 3 above)
Comment (KirillPamPam) [reply src/main/kotlin/io/emeraldpay/dshackle/auth/service/KeyReader.kt]: Fixed

File: src/main/kotlin/io/emeraldpay/dshackle/auth/processor/AuthProcessor.kt (hunk 4 above)
Comment (KirillPamPam) [reply src/main/kotlin/io/emeraldpay/dshackle/auth/processor/AuthProcessor.kt]: Agree, fixed

**Review by a10zn8 — APPROVED** [review]

**Review by Termina1 — APPROVED** [review]

**Pull request merged**